garage/tf/algos/td3.py

```diff
@@ -0,0 +1,252 @@
+            qval = self.qf.get_qval_sym(
+                obs, next_action, name='policy_action_qval')
+            q2val = self.qf2.get_qval_sym(
+                obs, next_action, name='policy_action_q2val')
+            next_qval = tf.minimum(qval, q2val)
+            with tf.name_scope('action_loss'):
+                action_loss = -tf.reduce_mean(next_qval)
+                if self.policy_weight_decay > 0.:
+                    policy_reg = tc.layers.apply_regularization(
+                        tc.layers.l2_regularizer(self.policy_weight_decay),
+                        weights_list=self.policy.get_regularizable_vars())
+                    action_loss += policy_reg
+
+            with tf.name_scope('minimize_action_loss'):
+                policy_train_op = self.policy_optimizer(
+                    self.policy_lr, name='PolicyOptimizer').minimize(
+                        action_loss, var_list=self.policy.get_trainable_vars())
+
+            f_train_policy = tensor_utils.compile_function(
+                inputs=[obs], outputs=[policy_train_op, action_loss])
+
+            # Set up qf training function
+            qval = self.qf.get_qval_sym(obs, actions, name='q_value')
+            q2val = self.qf2.get_qval_sym(obs, actions, name='q2_value')
+            with tf.name_scope('qval_loss'):
+                qval_loss = (tf.reduce_mean(tf.squared_difference(y, qval)) +
+                             tf.reduce_mean(tf.squared_difference(y, q2val)))
+                if self.qf_weight_decay > 0.:
+                    qf_reg = tc.layers.apply_regularization(
+                        tc.layers.l2_regularizer(self.qf_weight_decay),
+                        weights_list=self.qf.get_regularizable_vars())
+                    qval_loss += qf_reg
+
+            with tf.name_scope('minimize_qf_loss'):
+                qf_train_op = self.qf_optimizer(
+                    self.qf_lr, name='QFunctionOptimizer').minimize(
+                        qval_loss, var_list=self.qf.get_trainable_vars())
+                qf2_train_op = self.qf_optimizer(
+                    self.qf_lr, name='QFunctionOptimizer').minimize(
+                        qval_loss, var_list=self.qf2.get_trainable_vars())
+
+            f_train_qf = tensor_utils.compile_function(
+                inputs=[y, obs, actions],
+                outputs=[qf_train_op, qval_loss, qval])
+            f_train_qf2 = tensor_utils.compile_function(
+                inputs=[y, obs, actions],
+                outputs=[qf2_train_op, qval_loss, q2val])
+
+            self.f_train_policy = f_train_policy
+            self.f_train_qf = f_train_qf
+            self.f_init_target = f_init_target
+            self.f_update_target = f_update_target
+            self.f_train_qf2 = f_train_qf2
+            self.f_init_target2 = f_init_target2
+            self.f_update_target2 = f_update_target2
+
+    @overrides
+    def optimize_policy(self, itr, samples_data):
```

Comment:
does this use the new runner interface?

Reply:
TD3 inherits the `train_once()` from DDPG.

Reply:
ok